bpf/nat/connecttime.go

```diff
@@ -84,50 +85,61 @@ func RemoveConnectTimeLoadBalancer(cgroupv2 string) error {
 	return nil
 }
 
-func installProgram(name, ipver, bpfMount, cgroupPath, logLevel string, maps ...bpf.Map) error {
+func installProgram(name, ipver, bpfMount, cgroupPath, logLevel string) error {
 
 	progPinDir := path.Join(bpfMount, "calico_connect4")
 	_ = os.RemoveAll(progPinDir)
 
 	var filename string
-
 	if ipver == "6" {
 		filename = path.Join(bpf.ObjectDir, ProgFileName(logLevel, 6))
 	} else {
 		filename = path.Join(bpf.ObjectDir, ProgFileName(logLevel, 4))
 	}
-	args := []string{"prog", "loadall", filename, progPinDir, "type", "cgroup/" + name + ipver}
-	for _, m := range maps {
-		args = append(args, "map", "name", m.GetName(), "pinned", m.Path())
-	}
 
-	cmd := exec.Command("bpftool", args...)
-	log.WithField("args", cmd.Args).Info("About to run bpftool")
+	// nolint
 	progName := "calico_" + name + "_v" + ipver
-	out, err := cmd.CombinedOutput()
+
+	log.WithField("filename", filename).Debug("Loading object file")
+	obj, err := libbpf.OpenObject(filename)
 	if err != nil {
-		err = errors.Wrapf(err, "failed to load program %s", progName)
-		goto out
+		return fmt.Errorf("failed to load program %s from %s: %w", progName, filename, err)
 	}
+	defer obj.Close()
 
-	cmd = exec.Command("bpftool", "cgroup", "attach", cgroupPath,
-		name+ipver, "pinned", path.Join(progPinDir, progName))
-	log.WithField("args", cmd.Args).Info("About to run bpftool")
-	out, err = cmd.CombinedOutput()
-	if err != nil {
-		err = errors.Wrapf(err, "failed to attach program %s", progName)
-		goto out
+	baseDir := "/sys/fs/bpf/tc/globals/"
+	for m, err := obj.FirstMap(); m != nil && err == nil; m, err = m.NextMap() {
+		// In case of global variables, libbpf creates an internal map <prog_name>.rodata
+		// The values are read only for the BPF programs, but can be set to a value from
+		// userspace before the program is loaded.
+		if m.IsMapInternal() {
+			log.WithField("prog", progName).Warn("Load-time configuration not supported, using defaults.")
```

Comment:
When do we hit this?

Reply:
When the program has globals. I am adding the globals in a follow up PR

Reply:
This is the follow up https://github.com/projectcalico/felix/pull/3046